go-client/client/session_stub.go

```diff
@@ -0,0 +1,191 @@
+func (tk *tokenResp) getToken() ([]byte, error) {
+	tk.Lock.Lock()
+	defer tk.Lock.Unlock()
+	if tk.Error != nil {
+		return nil, tk.Error
+	} else {
+		token := make([]byte, len(tk.Token))
+		copy(token, tk.Token)
+		return token, nil
+	}
+}
+
+// setToken thread-safely sets the session token to a new value.
+func (tk *tokenResp) setToken(tok []byte) {
+	tk.Lock.Lock()
+	tk.Token = tok
+	tk.Lock.Unlock()
+}
+
+// setError thread-safely sets an error value for the session token.
+func (tk *tokenResp) setError(err error) {
+	tk.Lock.Lock()
+	tk.Error = err
+	tk.Lock.Unlock()
+}
+
+// A refresher stores the current client token and sends periodic keepalive messages to refresh the client token.
+type refresher struct {
+	ctx         context.Context
+	sessionStub sessionpb2.SessionServiceClient
+
+	token *tokenResp // the actual client token, which gets periodically updated.
+
+	cancelCh <-chan struct{} // if this channel closes, the refresher should stop.
+
+	timeoutMillis int64 // time (in milliseconds) before the token should be refreshed again.
+}
+
+// refreshLoop is an endless loop that will update the token when necessary.
+// It can be canceled by closing the cancelCh channel.
+func (ref *refresher) refreshLoop() {
+	for {
+		select {
+		case <-ref.cancelCh:
+			return
+		case <-time.After(time.Duration(ref.timeoutMillis) * time.Millisecond):
+			err := ref.refresh()
+			if err != nil {
+				return
+			}
+		}
+	}
+}
+
+// startRefresher begins a background goroutine that will update the passed token whenever it would time out.
+func startRefresher(ctx context.Context, sessionStub sessionpb2.SessionServiceClient, token *tokenResp, cancelCh <-chan struct{}) error {
+	handshakeReq := &sessionpb2.HandshakeRequest{AuthProtocol: 1, Payload: [](byte)("hello godeephaven")}
+	handshakeResp, err := sessionStub.NewSession(ctx, handshakeReq)
+	if err != nil {
+		return err
+	}
+
+	ctx = metadata.NewOutgoingContext(context.Background(), metadata.Pairs("deephaven_session_id", string(handshakeResp.SessionToken)))
+
+	token.setToken(handshakeResp.SessionToken)
+
+	timeoutMillis := handshakeResp.TokenExpirationDelayMillis / 2
+
+	ref := refresher{
+		ctx:         ctx,
+		sessionStub: sessionStub,
+
+		token: token,
+
+		cancelCh: cancelCh,
+
+		timeoutMillis: timeoutMillis,
+	}
+
+	go ref.refreshLoop()
+
+	return nil
+}
+
+// refresh is what actually refreshes the client token. It makes a RefreshSessionToken request,
+// and then updates the token struct.
+func (ref *refresher) refresh() error {
+	oldToken, err := ref.token.getToken()
+	if err != nil {
+		return err
+	}
+
+	handshakeReq := &sessionpb2.HandshakeRequest{AuthProtocol: 0, Payload: oldToken}
+	handshakeResp, err := ref.sessionStub.RefreshSessionToken(ref.ctx, handshakeReq)
+
+	if err != nil {
+		ref.token.setError(err)
+		return err
+	} else {
+		ref.token.setToken(handshakeResp.SessionToken)
+	}
+
+	ref.timeoutMillis = handshakeResp.TokenExpirationDelayMillis / 2
```

Comment:
Any thread safety concerns with this assignment? 

If the read and the write are not in the same thread, there is no guarantee that other threads will see it in a timely way.

In general, it is much easier to guarantee correct concurrent behavior if the entire algo uses channels.  I suspect this whole file can be written in a more concise and easier to verify correct way using channels.

See https://go.dev/ref/mem#tmp_2 to understand some of the problems.

Reply:
There are only two variables that are shared between threads here: `tokenResp` (which is already protected by a mutex) and `cancelCh` (which is a channel). Since every other variable only exists in a single goroutine, all other reads/writes are safe.

I will continue thinking about how I could write this using channels.